### PR TITLE
Test partitioning

### PR DESCRIPTION
**Pull request by fedepaol**
This tests the pf partitioning feature.
Also, it adds some utility functions that checks that the node network state(s) are stable between tests.

### CODE REVIEW

**Comment by zenghui-shi** [comment]
/cc @pliurh 

I noticed there were SR-IOV Partitioning tests added by Peng under `e2e` tests.
My understanding is that both conformance and e2e tests are integration tests. We might want to think about converging the two together and not having them located in different suites.
Thoughts?

**Comment by fedepaol** [comment]
> /cc @pliurh
> 
> I noticed there were SR-IOV Partitioning tests added by Peng under `e2e` tests.
> My understanding is that both conformance and e2e tests are integration tests. We might want to think about converging the two together and not having them located in different suites.
> Thoughts?

My understanding for those tests was that they are shaped for a specific installation (for example, they tie to intel as the vendor).
But I agree that we should try to leverage what is already there. Would it make sense to move keep conformance as the "generic" one where we move (at least) those tests that are meant to belong to the test plan? Ideally, we should be able to run conformance against any cluster with at least one sriov enabled node. Another thing that may help in that regard is to validate them against the cluster that is used for the tests that are currently in place.

**Comment by pliurh** [comment]
> > /cc @pliurh
> > I noticed there were SR-IOV Partitioning tests added by Peng under `e2e` tests.
> > My understanding is that both conformance and e2e tests are integration tests. We might want to think about converging the two together and not having them located in different suites.
> > Thoughts?
> 
> My understanding for those tests was that they are shaped for a specific installation (for example, they tie to intel as the vendor).
> But I agree that we should try to leverage what is already there. Would it make sense to move keep conformance as the "generic" one where we move (at least) those tests that are meant to belong to the test plan? Ideally, we should be able to run conformance against any cluster with at least one sriov enabled node. Another thing that may help in that regard is to validate them against the cluster that is used for the tests that are currently in place.

Yes, all the tests under the e2e directory are only designed to run against our baremetal CI environment, as we need to draft the policy according to the hardware layout.

I think adding conformance tests is a good idea. Currently, the 'operator' directory contains the tests for the 'SriovNetwork' and 'SriovOperatorConfg' CR, which don't need SRIOV hardware at all; while the 'e2e' directory contains all the tests for the 'SriovNetworkNodePolicy' CR, and they are hardware relevant. You may want to take a look at the cases there.
Currently, the operator-sdk test framework is used to deploy the operator before each test suit. But I can separate the operator deployment from the case execution so that all the tests can be executed one after another.

**Comment by zenghui-shi** [comment]
> > > /cc @pliurh
> > > I noticed there were SR-IOV Partitioning tests added by Peng under `e2e` tests.
> > > My understanding is that both conformance and e2e tests are integration tests. We might want to think about converging the two together and not having them located in different suites.
> > > Thoughts?
> > 
> > 
> > My understanding for those tests was that they are shaped for a specific installation (for example, they tie to intel as the vendor).
> > But I agree that we should try to leverage what is already there. Would it make sense to move keep conformance as the "generic" one where we move (at least) those tests that are meant to belong to the test plan? Ideally, we should be able to run conformance against any cluster with at least one sriov enabled node. Another thing that may help in that regard is to validate them against the cluster that is used for the tests that are currently in place.
> 
> Yes, all the tests under the e2e directory are only designed to run against our baremetal CI environment, as we need to draft the policy according to the hardware layout.
> 
> I think adding conformance tests is a good idea. Currently, the 'operator' directory contains the tests for the 'SriovNetwork' and 'SriovOperatorConfg' CR, which don't need SRIOV hardware at all; while the 'e2e' directory contains all the tests for the 'SriovNetworkNodePolicy' CR, and they are hardware relevant. You may want to take a look at the cases there.
> Currently, the operator-sdk test framework is used to deploy the operator before each test suit. But I can separate the operator deployment from the case execution so that all the tests can be executed one after another.

so what we going to do are:
1) use conformance folder for generic cases that can be run against every SR-IOV enabled cluster
2) use e2e folder for adding environment specific config ( generic test may located in conformance, but environment specific policy config may put here ) or tests.
3) separate operator installation part from run-test.sh script into generic script.

**Comment by fedepaol** [comment]
> > > /cc @pliurh
> > > I noticed there were SR-IOV Partitioning tests added by Peng under `e2e` tests.
> > > My understanding is that both conformance and e2e tests are integration tests. We might want to think about converging the two together and not having them located in different suites.
> > > Thoughts?
> > 
> > 
> > My understanding for those tests was that they are shaped for a specific installation (for example, they tie to intel as the vendor).
> > But I agree that we should try to leverage what is already there. Would it make sense to move keep conformance as the "generic" one where we move (at least) those tests that are meant to belong to the test plan? Ideally, we should be able to run conformance against any cluster with at least one sriov enabled node. Another thing that may help in that regard is to validate them against the cluster that is used for the tests that are currently in place.
> 
> Yes, all the tests under the e2e directory are only designed to run against our baremetal CI environment, as we need to draft the policy according to the hardware layout.
> 
> I think adding conformance tests is a good idea. Currently, the 'operator' directory contains the tests for the 'SriovNetwork' and 'SriovOperatorConfg' CR, which don't need SRIOV hardware at all; while the 'e2e' directory contains all the tests for the 'SriovNetworkNodePolicy' CR, and they are hardware relevant. You may want to take a look at the cases there.
> Currently, the operator-sdk test framework is used to deploy the operator before each test suit. But I can separate the operator deployment from the case execution so that all the tests can be executed one after another.

I am not completely sure I understood what you are suggesting. 

So, my idea was to leverage the sriovnodenetworkstate to adapt those tests that needed configuration to what the cluster had to offer. In that regard, this PR adds a test that validates the partitioning without hardcoding the vendor or the pf name.
Now, I see some different approaches:
- for tests that do not rely on the cluster "structure" we can rely on the 'operator' tests, adapting them in case some details from the test plan is missing (after being able to run them without installing the operator, which shouldn't be hard)
- for tests that rely on the cluster structure (those in e2e), we can either have duplication, try to make our side generic and cover what's being tested on e2e or do the other way around, trying to make e2e generic as well. I am afraid of this one as it feels a bit too invasive and risks to break what's already in place.

File: conformance/operator_test.go [hunk 1]
```diff
@@ -16,19 +17,27 @@ import (
...
 		err := namespaces.Create(namespaces.Test, clients)
 		Expect(err).ToNot(HaveOccurred())
 
-		err = namespaces.Clean(namespaces.Test, clients)
+		sriovInfos, err = cluster.DiscoverSriov(clients, operatorNamespace)
 		Expect(err).ToNot(HaveOccurred())
 
-		// This also validates that sriov is properly configured, ignoring the configuration ATM
-		_, err = cluster.DiscoverSriov(clients, operatorNamespace)
+		err = namespaces.Clean(operatorNamespace, namespaces.Test, clients)
```
Comment (SchSeba) [inline conformance/operator_test.go]: I think we can remove this one you have the same one in the `BeforeEach`

File: conformance/operator_test.go [hunk 2]
```diff
@@ -85,6 +94,114 @@ var _ = Describe("operator", func() {
...
+						},
+						DeviceType: "netdevice",
+					},
+				}
+
+				err = clients.Create(context.Background(), firstConfig)
+				Expect(err).ToNot(HaveOccurred())
+
+				Eventually(func() sriovv1.Interfaces {
+					var nodeState *sriovv1.SriovNetworkNodeState
```
Comment (SchSeba) [inline conformance/operator_test.go]: can we remove this?

File: conformance/operator_test.go [hunk 3]
```diff
@@ -85,6 +94,114 @@ var _ = Describe("operator", func() {
...
+					nodeState, err = clients.SriovNetworkNodeStates(operatorNamespace).Get(node, metav1.GetOptions{})
+					return nodeState.Spec.Interfaces
+				}, 3*time.Minute, 1*time.Second).Should(ContainElement(MatchFields(
+					IgnoreExtras,
+					Fields{
+						"NumVfs": Equal(5),
```
Comment (SchSeba) [inline conformance/operator_test.go:176]: Can you add here the `inf.name`?

**Review by SchSeba — CHANGES_REQUESTED** [review]
Few comments

**Review by SchSeba — APPROVED** [review]
Great job!

**Pull request merged**